### ISSUE TITLE
Find a replacement for Karma

### ISSUE THREAD

**Issue by hudson-newey** [Allcharles/workbench-client#2139]
Karma has been deprecated since April 2023.

The Angular [has suggested Jest and Web Test Runner](https://blog.angular.dev/moving-angular-cli-to-jest-and-web-test-runner-ef85ef69ceca) as replacements for Karma.

We should investigate which option will provide:

- The smoothest transition
	- Are the APIs and configs compatible with our existing code base?
- The most stable CI experience going forwards
- The fastest development experience
	- How intuitive is the testing framework
	- Can add threaded parallelism by using a new test runner (reducing the downtime waiting for tests to finish)

**Comment by hudson-newey** [comment]
So reading more on our options:

We have a decision between Jest OR Web Test Runner (wtr).

- Jest would replace BOTH Karma and Jasmine
- While wtr would only replace Karma

---

- Jest uses a NodeJS environment to mock browser api's
- Wtr uses a real browser, meaning that it is more memory intensive but mocks a real environment

I'm thinking that wtr is a better solution because it only replaces Karma which is much less work, and I do not have any criticisms of Jasmine

---

However, I don't really see the need for wtr. When we could just go directly to playwright.

Wtr is not a test runner, but an adapter on top of playwright, selenium, etc...
But in my mind that is what the ng-cli is for...
This would mean that we'd have three script runner abstractions when we run a test.
e.g. package.json > ng-cli > wtr > playwright

**Comment by atruskie** [comment]
Agree on wtr,

And not needing the abstraction

**Comment by hudson-newey** [comment]
Karma alternatives have **much** better support in newer Angular versions. So we should probably upgrade Angular first

See: #2258

**Comment by hudson-newey** [comment]
It looks like the ng-cli wtr builder is still VERY experimental (even in newer versions of Angular).

![Image](https://github.com/user-attachments/assets/9af4f5a1-74c5-44aa-8110-b9590314e4e5)

It still doesn't support a wide range of features that the old Karma builder supported.

And the Jest builder seems to be even further behind.

It would be possible to forgo the ng-cli for tests, and go directly to Playwright.
But that's not "the Angular way" and I think it'd be a maintenance burden (more of a maintenance burden than just using Karma until the wtr ng-cli builder is stable).

---

Small rant: It's annoying that Karma was deprecated before a suitable (official and stable) replacement was created. Now we're stuck with a deprecated test runner with no official replacements.

(there are a few community made ng-cli builders, but I wish there was an official tool).

The Angular team still recommends using Karma in their official docs (<https://angular.dev/guide/testing>), so I'm reluctantly thinking that we hold out replacing Karma until the Angular team marks the wtr builder as stable.

**Comment by hudson-newey** [comment]
I'm going to double down on waiting til the Angular cli team has fully integrated a stable version of wtr into the angular cli.

Although Karma is deprecated, the Angular team is still recommending using Karma until wtr is stable enough for production and can be a drop in replacement for the Karma builder.

---

I have tried to use wtr directly, bypassing the ng cli, but I found that I was just re-implementing the same code that the ng cli team was writing (duplicating work).

Although the Angular team is lagging behind on making wtr stable (look at the blame for the wtr builder <https://github.com/angular/angular-cli/tree/main/packages/angular_devkit/build_angular/src/builders/web-test-runner>), I think it's still best to wait until the ng cli produce a stable version.

---

I tried using Jest as well, and the Jest builder seems to be more stable, but it lacks a lot of features that even Karma has (it doesn't run in a browser, so there's no browser debugger or inspect element, which makes tests harder to debug).
Additionally, I found that Jest was even heavier than Karma.

---

I could migrate to Playwright. The angular team actually supports (and recommends) playwright for e2e tests.

However, this would also rip out Jasmine, and I feel like that would be a job too large and a waste of time.
Especially when they are promising a drop-in replacement for Karma.

As a side note: I thought that wtr was just a wrapper for playwright. While it does support running wtr through playwright browsers, it does have its own native browsers which the angular team is using.

---

In conclusion:

- Karma is still the "recommended runner" for Angular apps (even though it's been deprecated for > 2 years)
- Web test runner is not remotely stable enough
- Manually adjusting web test runner to Angular has me repeating the ng-cli teams work
- Jest is too slow, and actually worse for our use case than just using Karma
- Migrating to Playwright is possible. But it means that we also has to replace Jasmine, which is a time pit that will not yield enough benefit (especially when the Angular team has promised that we don't have to do all of this work)

**Comment by atruskie** [comment]
Agree on all points, lets put this in the backlog

**Comment by hudson-newey** [comment]
New issue:

When serving an audio file through the Karma server (using the new `@angular-devkit/build-angular:karma` builder), the file served through the server is modified & corrupted the file after the test runs.

Note: Downloading the audio file through the server before the first test runs does not result in a modified/corrupted file.

---

We think this is because somewhere in the line, the audio file is being read/write as utf-8 because it has utf replacement characters.

I fixed this issue by using a base64 encoding of the tested audio file.

We should be able to revert this hacky change once Karma is replaced.

**Comment by hudson-newey** [comment]
The Angular team has landed on Vitest as the recommended replacement

We'll be able to replace Karma once we update to Angular 21

https://angular.dev/guide/testing

**Comment by atruskie** [comment]
I know you've answered before: but what are our blockers for angular 21?

**Comment by hudson-newey** [comment]
> I know you've answered before: but what are our blockers for angular 21?

There are no hard blockers except waiting for our dependencies to update to Angular 21

Updating the Angular 20 is proving more difficult than usual because ngx-datatable made a bunch of previously public api's private which broke a bunch of our test assertions that queried the fixture for these previously public components

(see the `public-api.ts` file changes) https://github.com/swimlane/ngx-datatable/commit/f352a4cd0c0d9c69c0f6195109f0c63c1a55abe2

---

I've cleaned up the tests and found that all of the references to these 4 now private APIs were for the most part not unique, so I was able to consolidate all of the assertions into some helper functions.

<img width="867" height="406" alt="Image" src="https://github.com/user-attachments/assets/74b484cc-550b-4487-bde7-52694a05f1c4" />

I was able to do a find & replace, so hopefully I just have to change these 4 functions to get the tests passing
